### ISSUE TITLE
Implement remaining `pycodestyle` rules

### ISSUE THREAD

**Issue by charliermarsh** [charliermarsh/ruff#2402]
- [ ] `E201` (`whitespace after '['`)
- [ ] `E202` (`whitespace before ')'`)
- [ ] `E203` (`whitespace before ','`)
- [ ] `E211` (`whitespace before '('`)
- [ ] `E221` (`multiple spaces before operator`)
- [ ] `E301` (`expected 1 blank line, found 0`)
- [ ] `E302` (`expected 2 blank lines, found 1`)
- [ ] `E303` (`too many blank lines (4)`)

**Comment by charliermarsh** [comment]
I'm happy to support these, but I likely won't enable them by default -- I'd like them to be opt-in.

**Comment by charliermarsh** [comment]
\cc @saadmk11

**Comment by charliermarsh** [comment]
First we need #1130.

**Comment by charliermarsh** [comment]
So far I have seven more rules done in #1130.

**Comment by pierrepo** [comment]
Thanks for this @charliermarsh 
I'm using a lot `pycodestyle` and `pydocstyle` with my students to teach them PEP8 and PEP257. Using `ruff` will be a great plus.

**Comment by charliermarsh** [comment]
#1130 is up to 14 rules. Just trying to knock out a few each days.

**Comment by charliermarsh** [comment]
Knocked out eight more rules behind the feature flag.

**Comment by charliermarsh** [comment]
Four more done.

**Comment by charliermarsh** [comment]
One more in #3225.

**Comment by charliermarsh** [comment]
One more in https://github.com/charliermarsh/ruff/pull/3249.

**Comment by charliermarsh** [comment]
One more in https://github.com/charliermarsh/ruff/pull/3344.

**Comment by kapilt** [comment]
rust newb, if someone wants to try this feature set, how do we build ruff with logical-line feature flag support and these style rules enabled?

**Comment by spaceone** [comment]
> rust newb, if someone wants to try this feature set, how do we build ruff with logical-line feature flag support and these style rules enabled?
@charliermarsh I would like to have a pip release for ruff with that special feature as pip/setup.py "package option" / "extra"/ "extras_require" enabled. So that you can call `pip install ruff[logical-line]` and have this enabled. Maybe it's possible to achieve that?

**Comment by charliermarsh** [comment]
I'm hoping that we can finish and release this soon enough that it's not worth shipping an `extra_requires` variant. (And if not, I'd probably rather ship with a command-line or feature flag rather than an `extras`, so that it's available regardless of how you install Ruff.)

**Comment by jamesbraza** [comment]
This conquest is awesome!  I see many of these as being checked (e.g. `E2` whitespace is done).

Just a friendly request that these gets added to the Rules in docs: https://beta.ruff.rs/docs/rules/#error-e

**Comment by calumy** [comment]
> Just a friendly request that these get eventually added to the Rules in docs: https://beta.ruff.rs/docs/rules/#error-e

I believe this will be added when #3689 lands, as documenting a rule before it has been enabled will potentially cause some confusion.

**Comment by hoel-bagard** [comment]
I'll take a look at E303.

**Comment by devbis** [comment]
E241 and E241 is missing in the list for pycodestyle.

E241  | multiple spaces after ','
E242 | tab after ','


Can you please add it to the checks?

**Comment by charliermarsh** [comment]
I believe those are intentionally omitted -- we're not planning to implement the rules that are omitted from pycodestyle's default configuration ("not unanimously accepted, and PEP 8 does not enforce them"), at least not to start.

**Comment by hoel-bagard** [comment]
I'm looking into `E122` (and I'll follow with the other `E1` rules after unless someones is working on them).

**Comment by MartinRoth** [comment]
Hello,

I just stumbled about a peculiar behaviour of E714. I had fixed the lint but forgot to remove the first not, i.e. I changed the first if statement to the second below and ruff (I am using ruff version v0.0.275) didn't complain anymore.

```
a = 8

if not a is None:
    print(a)

if not a is not None:
    print(a)
```

I think the lint should also check for the second version, or am I wrong?

**Comment by ksunden** [comment]
Gentle voice of support for E122 and E302 (the latter being included in #4694 which has been in draft for a while.) (ping @hoel-bagard, as both author of that PR and as the one who said you were looking into E122)

These are the last two rules from Matplotlib's flake8 config. Once available, I'd look into swapping over CI/pre-commit hooks to ruff (and then perhaps look into adding more checks, though for now have gone with "parity")

We do not use `black`, so these are checks which fall into the category of "black would just handle them".

**Comment by hoel-bagard** [comment]
@ksunden Sorry for taking a long time, for E302, I have started to see if I could find a way to handle comments with the tools currently present in ruff (otherwise a new tool/way of tracking tokens will need to be implemented as suggested in the PR). I was a bit busy recently, but I should be able to spend more time on it in the near future.

For E122, I've just edited my comment saying I was looking into it. A while ago I asked a question on the discord, but did not get an answer (and then I went back to working on the E302 rule and forgot about it). I could restart working on it assuming that having a token emitted for the backslash is fine, in which case it shouldn't take too much time.

> I'm looking into adding the E122 rule of pycodestyle (https://www.flake8rules.com/rules/E122.html). It's a rule that checks for indentation within continuation lines.


> Implementing it using logical lines seemed natural (and that's what pycodestyle does). It seems like it would work for continuation lines made using parentheses since a NonLogicalNewline is emitted for the new line.
> However, when using a backslash (case covered by pycodestyle), no NonLogicalNewline token is created, rendering it impossible to know if there is a newline within the logical line.
> 
> Could a NonLogicalNewline be emitted when there is a backslash, or does the rule need to be implemented using physical lines ?
> 
> Example with parentheses:
> ```python
> print("E122", (
> "dent"))
> ```
> 
> Example using a backslash:
> ```python
> if some_very_very_very_long_variable_name or var \
> or another_very_long_variable_name:
>     raise Exception()
> ```

**Comment by Daku24** [comment]
Is there an update on when E3 Blank line will get released?

**Comment by hoel-bagard** [comment]
I've been doing the `E122` rule. I'm at a point where it's working, but I need to clean things up and add the autofix. I'll go back to the E3 rule once that's done.

**Comment by Avasam** [comment]
`W391` is checked but I don't see the rule in Ruff

**Comment by hoel-bagard** [comment]
@charliermarsh Since #9266 got merged, the `E3` rules can be checked here.

**Comment by spaceone** [comment]
Are meanwhile all above checked rules now available, at least behind the preview flag?

**Comment by charliermarsh** [comment]
@spaceone - Yup, they're all behind `--preview`.

**Comment by kaddkaka** [comment]
When was ` E303 ("too many blank lines (3)")` added? I have some python code and this rule is not discovered by ruff.

Example:

```py
A = 100



B = 100
```

**Comment by hoel-bagard** [comment]
@buhtz Have you selected the E303 rule ? For example with:

```console
ruff foo.py --preview --select E3
```

**Comment by buhtz** [comment]
> @buhtz Have you selected the E303 rule ? For example with:
> 
> ```
> ruff foo.py --preview --select E3
> ```

No, I have not. As [discussed here](https://github.com/astral-sh/ruff/discussions/10153#discussioncomment-8622602) and how I understand [the docs](https://docs.astral.sh/ruff/preview/#enabling-preview-mode) the `--preview` flag to select all "preview" (unstable, new) rules.

**Comment by baco** [comment]
> > @buhtz Have you selected the E303 rule ? For example with:
> > ```
> > ruff foo.py --preview --select E3
> > ```
> 
> No, I have not. As [discussed here](https://github.com/astral-sh/ruff/discussions/10153#discussioncomment-8622602) and how I understand [the docs](https://docs.astral.sh/ruff/preview/#enabling-preview-mode) the `--preview` flag to select all "preview" (unstable, new) rules.

Well... that's the issue. As for now, all the `E3` rules have just arrived and are on their testing period, hence, the only way to try them is with the `--preview` flag.

You can see that here, in the [Rules](https://docs.astral.sh/ruff/rules/), and see that all the rules available under the `--preview` are marked with 🧪 (a test tube emoji). And the rules you are looking for:
![image](https://github.com/astral-sh/ruff/assets/102958/5c2bb2c1-d281-438b-aa65-5035b6d437da)
all have the test tube emoji

**Comment by buhtz** [comment]
> the only way to try them is with the `--preview` flag.

No, you are wrong. That flag alone does not help. You have to combine `--preview` with `--select`. That is the missing key fact here. I know you know it. But some readers might misunderstand that. Wording is very important here.

**Comment by baco** [comment]
Ahhh, ok, ok, that's right, the combination of both flags. `--preview` doesn't turn on all the “unstable rules”. The flag rather stops filtering out rules matching the “selection rules” (if that's a thing) that happen to be unstable.

Is that what you are pointing out here?

**Comment by buhtz** [comment]
> Is that what you are pointing out here?

Exactly.

**Comment by kaddkaka** [comment]
> Note: some of the checked-off rules are still gated behind the `logical_lines` feature flag. To see the list of rules enabled in the current release, refer to the [docs](https://github.com/charliermarsh/ruff#error-e).

Is this still true? Is there a better way than cross referencing https://docs.astral.sh/ruff/rules/#pycodestyle-e-w with the list in this issue?

Where can I find info about "logical_lines"? A search in the docs gave 0 matches.

**Comment by MichaReiser** [comment]
> Note: some of the checked-off rules are still gated behind the logical_lines feature flag. To see the list of rules enabled in the current release, refer to the [docs](https://github.com/charliermarsh/ruff#error-e).

No, that's no longer the case. You can opt in by using `--preview` and enabling the desired rules (e.g. `--extend-select=E3`) 

All checked rules in this issue are implemented as stable or preview rules and the rules use the same codes as `pycodestyle`. So there should be no need to cross-reference except if you need to know if the rule is preview only.

**Comment by Daku24** [comment]
Anyone know how to use the preview E3 Blank line rules in vscode with the ruff extension. I have the preview version installed in vscode.

**Comment by dhruvmanila** [comment]
@Daku24 By preview, we mean https://docs.astral.sh/ruff/preview/, you don't need to have the _preview_ version of ruff-vscode extension. If you have a local config for Ruff, you can enable preview and include `E3`:

```toml
[tool.ruff.lint]
preview = true
extend-select = ["E3"]
```

If you don't have a Ruff config file or can't create or edit it, you can use [`lint.args`](https://github.com/astral-sh/ruff-vscode#settings).

I hope this helps :)

**Comment by Alex-ley-scrub** [comment]
This is awesome work 👏 

Once E11, E12, and E13 rules are added with automatic fixes, if we run something like this:
```
ruff check foo.py --fix --preview --select E1
```
would we would get the ~same result as something like this?:
```
autopep8 foo.py --in-place [--aggressive] 
```
If so, that would be amazing! autopep8 is really good at fixing broken indentation and adding this to ruff would be 😍 

E1 rules don't have automatic fixes at the minute - is that just a case of priority/time or is there a technical challenge / performance concern / reluctance for this to be default behaviour etc. as somewhat discussed in https://github.com/astral-sh/ruff/pull/1130

**Comment by kaddkaka** [comment]
Most of the pycodestyle rules are in preview set. When will they be upgraded to stable rules?

**Comment by kaddkaka** [comment]
I don't think that should be a reason to not promote those rules to stable if they are acting correctly.

Details:
1. We have a lot of files that we are not formatting with ruff.
2. Turning on `--preview` is unwanted since that together with `PL` would enable other rules like PLC1901 (see https://github.com/astral-sh/ruff/issues/14602)

**Comment by beauxq** [comment]
In this list, E704 has a check mark,
but E704 is not in the list here https://docs.astral.sh/ruff/rules/#error-e
What's going with that rule?

**Comment by MichaReiser** [comment]
> I believe the concerns around the ignore list have been resolved, so we could bring it back. I'll remove the check mark for now.

Can you expand on which concerns were addressed? I think the concern that it isn't an *on-by-default* rule still applies

**Comment by ntBre** [comment]
> Can you expand on which concerns were addressed? I think the concern that it isn't an _on-by-default_ rule still applies

Oh I assumed that this just meant a way to disable rules and that this didn't exist at the time. I forgot the E rules are on by default.

**Comment by beauxq** [comment]
If we don't want it as "E" because "E" is too default, maybe we could have the rule somewhere else, like "RUF"?

But I think the better way to go would be to keep it in "E", but match black's treatment allowing the `...` for `.pyi` and overloads (which should be done even if it is moved somewhere else, like "RUF").
That way, I think people wouldn't mind it being on by default.

**Comment by kaddkaka** [comment]
> I don't think that should be a reason to not promote those rules to stable if they are acting correctly.
> 
> Details:
> 
> 1. We have a lot of files that we are not formatting with ruff.
> 2. Turning on `--preview` is unwanted since that together with `PL` would enable other unwanted rules like PLC1901 (see [How to deal with counter-argument to rules? Concrete example PLC1901 #14602](https://github.com/astral-sh/ruff/issues/14602))

@MichaReiser you presented an argument against stabilizing these rules as they are already covered by ruff format. (https://github.com/astral-sh/ruff/discussions/14331#discussioncomment-11263890) 

I would like to run just the linter part without the formatter. If that is not intended usecase of ruff and these rules will not be stabilized we will stick with other linters. (I would prefer just one, ruff 😇) 

Any chance of re-evaluation? Or some middle-ground like bringing the rules into stable but making them "more" opt-in by grouping them in a subcategory of E? (I guess it's quite common to enable the E rules which means just stabilizing them would make them appear in many people's lint setup automatically. Maybe that's not a problem though.)

**Comment by jgbishop** [comment]
Are E122 - E131 still planned?

**Comment by hoel-bagard** [comment]
There is [a PR](https://github.com/astral-sh/ruff/pull/13585) implementing the E12X rules.

**Comment by kaddkaka** [comment]
> > I don't think that should be a reason to not promote those rules to stable if they are acting correctly.
> > Details:
> > 
> > 1. We have a lot of files that we are not formatting with ruff.
> > 2. Turning on `--preview` is unwanted since that together with `PL` would enable other unwanted rules like PLC1901 (see [How to deal with counter-argument to rules? Concrete example PLC1901 #14602](https://github.com/astral-sh/ruff/issues/14602))
> 
> [@MichaReiser](https://github.com/MichaReiser) you presented an argument against stabilizing these rules as they are already covered by ruff format. ([#14331 (comment)](https://github.com/astral-sh/ruff/discussions/14331#discussioncomment-11263890))
> 
> I would like to run just the linter part without the formatter. If that is not intended usecase of ruff and these rules will not be stabilized we will stick with other linters. (I would prefer just one, ruff 😇)
> 
> Any chance of re-evaluation? Or some middle-ground like bringing the rules into stable but making them "more" opt-in by grouping them in a subcategory of E? (I guess it's quite common to enable the E rules which means just stabilizing them would make them appear in many people's lint setup automatically. Maybe that's not a problem though.)

Over half a year has passed, could we discuss this?

**Comment by NN---** [comment]
I tried to move to ruff but found that not all flake8 rules are applied in ruff.
So I probably need both tools ?

**Comment by buhtz** [comment]
I am not related to astral/ruff. But as upstream maintainer I do combine several tools because of the issues you observed.

1. Ruff - Because it is fast.
2. Flake8 - Covers some rules Ruff don't.
3. PyLint - Very slow, but damn "pedantic", not only about coding style but about code smells.
4. Additionally I add `reuse lint` and `codespell` at the end in most of my projects.

e.g. See this for a real world example where I add all this linters as unit tests. https://github.com/bit-team/backintime/blob/dev/common/test/test_lint.py

**Comment by NN---** [comment]
Is there a point to run ruff if you still run flake8?
Yes it is super fast but doesn't have some important rules.

**Comment by buhtz** [comment]
> Is there a point to run ruff if you still run flake8?

Yes, it definitely is, because Ruff does not implement all Flake8 rules. But they are working on it. In my projects it is very rare that Flake8 (after running ruff) give some errors. In my foggy memory I would say it is often PEP8 coding style stuff, like to much blank lines or something like this.

But it is also the other way around. Ruff catch some errors that Flake8 won't.

**Comment by xmatthias** [comment]
> But it is also the other way around. Ruff catch some errors that Flake8 won't.

Exactly this! 

> Yes, it definitely is, because Ruff does not implement all Flake8 rules. 

But it implements most rules (at least I personally) often violate - so having `ruff check` pass is almost always sufficient.
Flake8 is still used for whitespace issues - but once this issue resolves - i expect to remove flake8 from my projects.

**Comment by 153957** [comment]
Wouldnt `ruff format` handle most of your whitespace issues?

**Comment by timj** [comment]
> Wouldnt ruff format handle most of your whitespace issues?

That was the conclusion I came to. `ruff format` makes all these missing checks irrelevant because you know that they can't ever happen.

**Comment by kaddkaka** [comment]
> > Wouldnt ruff format handle most of your whitespace issues?
> 
> That was the conclusion I came to. `ruff format` makes all these missing checks irrelevant because you know that they can't ever happen.

But that's irrelevant as `ruff format` is not `ruff check` and everyone running linters might not want to run formatters. This has already been up several times.

**Comment by timj** [comment]
I understand that check is not format. I've been following this GitHub issue a very long time hoping that the checks would turn up at some point for code repos where I'm not allowed to use `format`. I was noting that in some cases it is thankfully irrelevant for me.

**Comment by pebaha-** [comment]
> > Wouldnt ruff format handle most of your whitespace issues?
> 
> That was the conclusion I came to. `ruff format` makes all these missing checks irrelevant because you know that they can't ever happen.

`ruff format` and `ruff check` are seperate commands and serve different purposes. If you use `ruff format` you also buy into black-compatible formatting, which not all projects subscribe to.

**Comment by buhtz** [comment]
Yes, `ruff format` is not linting but auto-formating. I am not a friend of that (or `black`).
You won't learn writing PEP8-conform code when a tool does make the job for you. It is always a good advise not to use things like that.

**Comment by MichaReiser** [comment]
Please avoid off-topic conversations on this issue. There are many users subscribed to it and each comment notifies all of them. Instead, open a new issue or discussion specific about your concern.

**Comment by spaceone** [comment]
Years passed. Is there a plan when these rules get into un-preview?
It really would help to get rid of slow pre-commit hooks, which are currently required.

**Comment by ntBre** [comment]
Yes, the new `formatting` category introduced as part of #1774 is almost entirely intended for the formatting-related pycodestyle rules. That gives us a place to stabilize these rules without mixing the formatting `E` rules into the correctness-related ones, which was our previous hesitation.
